core/src/km_core_state_api.cpp

```diff
@@ -363,4 +364,35 @@ km_core_cp * km_core_state_context_debug(
   result[s.size()] = 0;
 
   return result;
-}
\ No newline at end of file
+}
+
+static bool
+state_has_action_type(km_core_state *state, uint8_t type) {
+  return std::any_of(
+      state->actions().begin(), state->actions().end(),
+        [type](const km::core::action &a) { return a.type == type; });
+}
+
+bool
+state_should_invalidate_context(km_core_state *state,
+                     km_core_virtual_key vk,
+                     uint16_t modifier_state,
+                     uint8_t is_key_down,
+                     uint16_t _kmn_unused(event_flags)) {
+  if (!is_key_down) {
+    return false;  // don't invalidate on keyup
+  }
+  // if emit_keystroke is present, check if a context reset is needed
+  if (state_has_action_type(state, KM_CORE_IT_EMIT_KEYSTROKE)) {
+    if (
+        // always reset on backspace
```

Comment:
Some more explanation here would be helpful.

```suggestion
        // when a backspace keystroke is emitted, it is because we are at the start of 
        // context, and we want to give the application the chance to process it, e.g. 
        // by moving to previous field. Note that context manipulation does not result 
        // in an emit_keystroke backspace action, as this is handled through the 
        // `code_points_to_delete` field. So we always invalidate context when a 
        // processor emits a backspace.
```